### PR TITLE
[web] Updates package:web dependency to ^0.5.0.

### PR DESCRIPTION
**Pull request by kevmoo**
This PR updates the packages that are using `package:web` to version `^0.5.0`.

Ancilliary changes:

* Bump `environment` to `flutter: ">=3.19.0"` and `sdk: ^3.3.0`.
* Bump version to next `Y`
* Clean-up code that was kept for compatibility with versions of `web: <0.5.0`.

The main exception to this is `package:google_sign_in_web`, which depends on a version of `google_identity_services_web` that has a dependency on package:web that is `<0.5.0`, so that package needs to have a range until `google_identity_services_web` gets published with the new ^0.5.0 dependency.

Co-Authored-By: David Iglesias<ditman@gmail.com>

### CODE REVIEW

**Comment by ditman** [comment]
I'll take over landing this one!

File: packages/web_benchmarks/lib/client.dart
Comment (ditman) [reply packages/web_benchmarks/lib/client.dart]: Removed, the rest of the occurrences are in `mock` or _test files, not in user code.

Comment (ditman) [reply packages/web_benchmarks/lib/client.dart]:
```console
$ grep deprecated_member_use * -Irn | grep "_web" | grep "dart:" | cut -d: -f1 | sort | uniq -c | sort -r
      8 shared_preferences/shared_preferences_web/example/integration_test/shared_preferences_web_test.dart
      4 image_picker/image_picker_for_web/example/integration_test/image_picker_for_web_test.dart
      2 webview_flutter/webview_flutter_web/test/web_webview_controller_test.mocks.dart
      2 webview_flutter/webview_flutter_web/test/legacy/webview_flutter_web_test.mocks.dart
      2 google_sign_in/google_sign_in_web/example/integration_test/web_only_test.mocks.dart
      2 google_sign_in/google_sign_in_web/example/integration_test/google_sign_in_web_test.mocks.dart
      2 google_maps_flutter/google_maps_flutter_web/example/integration_test/overlays_test.mocks.dart
      2 google_maps_flutter/google_maps_flutter_web/example/integration_test/google_maps_plugin_test.mocks.dart
      2 google_maps_flutter/google_maps_flutter_web/example/integration_test/google_maps_controller_test.mocks.dart
```

Do we want to remove these from `_test.dart` files as well?

Comment (stuartmorgan) [reply packages/web_benchmarks/lib/client.dart]: Autogenerated mocks are fine; the `ignore`s there are also auto-generated so not something we maintaine.

The two test files there are specific exceptions called out in the new gardener docs. They are annotating places where we have integration tests of our own deprecated methods (integration tests are technically a different bundle, so trip that check instead of `deprecated_member_use_from_same_package`), and we don't track those as tech debt since they are not actually cross-package, and will be removed in the same PR that removes the deprecated APIs themselves. Since we don't want to track them, we don't want them to show up in the manual audit, so we `ignore` them.

**Comment by ditman** [comment]
```console
$ grep js_util * -Irn | grep "_web" | grep "dart:" | cut -d: -f1 | sort | uniq -c | sort -r
      5 camera/camera_web/example/integration_test/camera_service_test.dart
      4 camera/camera_web/lib/src/shims/dart_js_util.dart
      3 camera/camera_web/lib/src/camera_service.dart
      1 google_maps_flutter/google_maps_flutter_web/lib/google_maps_flutter_web.dart
      1 google_maps_flutter/google_maps_flutter_web/example/integration_test/google_maps_plugin_test.dart
      1 camera/camera_web/example/integration_test/helpers/mocks.dart
```

js_util seems to be almost gone from the repo (except from known non-migrated packages)

**Review by ditman — COMMENTED** [review]
~~I think this is ready to land, PTAL @stuartmorgan!~~

File: packages/web_benchmarks/lib/client.dart
Comment (ditman) [reply packages/web_benchmarks/lib/client.dart]: Ah, fantastic! I totally missed the gardener docs about those two files 😅

I'll mark this conversation as resolved then!

**Comment by ditman** [comment]
I think this is OK now. PTAL @stuartmorgan (and @kevmoo)!

**Comment by kevmoo** [comment]
I'd review it if I could! Thanks @ditman

File: packages/google_identity_services_web/lib/src/js_interop/package_web_tweaks.dart
Comment (stuartmorgan) [inline packages/google_identity_services_web/lib/src/js_interop/package_web_tweaks.dart:13]: Why are we using extensions? https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-extension

File: packages/google_sign_in/google_sign_in_web/CHANGELOG.md
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in_web/CHANGELOG.md:3]: This seems like an odd description since it actually does the update.

(Also, nit: this package is all web code, so doesn't need to specify that here.)

File: packages/google_sign_in/google_sign_in_web/pubspec.yaml
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in_web/pubspec.yaml:28]: Is this comment explaining the lower limit or the upper limit? More details would help.

File: packages/metrics_center/pubspec.yaml
Comment (stuartmorgan) [inline packages/metrics_center/pubspec.yaml:15]: Wow, this is Pigeon-level version revving 😁

File: packages/video_player/video_player_web/example/integration_test/utils.dart
Comment (stuartmorgan) [inline packages/video_player/video_player_web/example/integration_test/utils.dart:22]: Same question here.

File: packages/metrics_center/pubspec.yaml
Comment (kevmoo) [reply packages/metrics_center/pubspec.yaml:15]: We could honestly do `^12.0.0` I think.

Comment (stuartmorgan) [reply packages/metrics_center/pubspec.yaml:15]: Normally I'd say we should favor keeping the broad compatibility, but since no packages depend on `metrics_center` it probably doesn't matter either way.

File: packages/google_sign_in/google_sign_in_web/pubspec.yaml
Comment (kevmoo) [reply packages/google_sign_in/google_sign_in_web/pubspec.yaml:28]: I *think* this is because `google_identity_services` has not been upgraded yet...

File: packages/google_identity_services_web/lib/src/js_interop/package_web_tweaks.dart
Comment (ditman) [reply packages/google_identity_services_web/lib/src/js_interop/package_web_tweaks.dart:13]: Extension types and extensions seem to be The New Way of doing JS-interop, rather than the `@staticInterop` annotation. See this: https://github.com/flutter/flutter/pull/143274 (I followed @kevmoo's lead here)

File: packages/google_sign_in/google_sign_in_web/pubspec.yaml
Comment (ditman) [reply packages/google_sign_in/google_sign_in_web/pubspec.yaml:28]: The comment explains the upper limit, this is the package I'll immediately update after landing this PR, so the upper limit will live there only for one small publish.

File: packages/video_player/video_player_web/example/integration_test/utils.dart
Comment (ditman) [reply packages/video_player/video_player_web/example/integration_test/utils.dart:22]: (Same answer as above)

File: packages/google_identity_services_web/lib/src/js_interop/package_web_tweaks.dart
Comment (stuartmorgan) [reply packages/google_identity_services_web/lib/src/js_interop/package_web_tweaks.dart:13]: Okay, if Hixie is on board with this use case an an exception to the Flutter style guide that's fine.

Comment (kevmoo) [reply packages/google_identity_services_web/lib/src/js_interop/package_web_tweaks.dart:13]: Should we ask @Hixie directly?

Comment (stuartmorgan) [reply packages/google_identity_services_web/lib/src/js_interop/package_web_tweaks.dart:13]: Most of the file is the same extension type pattern that he just approved in the other PR, so I was taking that as confirmation that it should be fine.

**Review by stuartmorgan — APPROVED** [review]
LGTM modulo the changelog nits

File: packages/google_sign_in/google_sign_in_web/CHANGELOG.md
Comment (ditman) [reply packages/google_sign_in/google_sign_in_web/CHANGELOG.md:3]: Agh shoot, I did a grammar pass for this description in other packages, but this one wasn't an exact match and I forgot. Thanks for flagging this!

File: packages/google_identity_services_web/lib/src/js_interop/package_web_tweaks.dart
Comment (ditman) [reply packages/google_identity_services_web/lib/src/js_interop/package_web_tweaks.dart:13]: > (This MAY be rewritable as extension type NullableTrustedTypesGetter._(web.Window) implements web.Window... (I'm going to try!))

OK, I tried:

```dart
extension type NullableTrustedTypesGetter._(web.Window _) implements web.Window {
```

instead of

```dart
extension NullableTrustedTypesGetter on web.Window {
```

And the compiler seems to take it, HOWEVER, when using the `web.window` instance of the `web.Window` class, my "wrapper" is not automagically applied, and I cannot access the nullable setter (because it should be of type `NullableTrustedTypesGetter`). This could be fixed by casting `as NullableTrustedTypesGetter` but... IMO that makes the code less legible.

So the rule of thumb seems to be:

* My custom JS-Interop types need to `extension type ... implements JSObject` (as described in the docs, this is [how `class Window` is defined in `package:web`](https://github.com/dart-lang/web/blob/975e55cbe4ac683c33d5256a91f4133fa97b4c34/lib/src/dom/html.dart#L2707-L2709))
* Existing JS-interop types that come from elsewhere and need some "tweaks" (like `web.Window` in this example) need to be a normal `extension ... on pkg.ForeignType`, so it gets applied automatically without us having to cast or doing anything to that foreign object.

(Of course, the rule of thumb will break down when we discover more complex use-cases, so take it with a grain of salt!)

File: packages/metrics_center/pubspec.yaml
Comment (ditman) [reply packages/metrics_center/pubspec.yaml:15]: Let's do it then!

File: packages/google_sign_in/google_sign_in_web/CHANGELOG.md
Comment (ditman) [reply packages/google_sign_in/google_sign_in_web/CHANGELOG.md:3]: I also made this small dependency change a `+Z` change; I'll bump `+Y` when the migration of `package:web` actually happens.

File: packages/metrics_center/pubspec.yaml
Comment (ditman) [reply packages/metrics_center/pubspec.yaml:15]: Bumped to ^12.0.0, see what happens.

**Pull request merged**
